File: conformance/tests/post-quantum-slhdsa/post-quantum-slhdsa.runar.ts
import { SmartContract, assert, verifySLHDSA_SHA2_128s } from 'runar-lang';
import type { ByteString } from 'runar-lang';

class PostQuantumSLHDSA extends SmartContract {
  readonly pubkey: ByteString;

  constructor(pubkey: ByteString) {
    super(pubkey);
    this.pubkey = pubkey;
  }

  public spend(msg: ByteString, sig: ByteString) {
    assert(verifySLHDSA_SHA2_128s(msg, sig, this.pubkey));
  }
}

File: conformance/tests/post-quantum-slhdsa/post-quantum-slhdsa.runar.rs
use runar::prelude::*;

#[runar::contract]
struct PostQuantumSLHDSA {
    #[readonly]
    pubkey: ByteString,
}

#[runar::methods(PostQuantumSLHDSA)]
impl PostQuantumSLHDSA {
    #[public]
    fn spend(&self, msg: &ByteString, sig: &ByteString) {
        assert!(verify_slh_dsa_sha2_128s(msg, sig, &self.pubkey));
    }
}
